Changes in module 'nf-core/savana/run'
--- modules/nf-core/savana/run/main.nf
+++ modules/nf-core/savana/run/main.nf
@@ -4,8 +4,8 @@
     conda "${moduleDir}/environment.yml"
     container "${workflow.containerEngine in ['singularity', 'apptainer'] && !task.ext.singularity_pull_docker_container
-        ? 'https://depot.galaxyproject.org/singularity/savana:1.3.7--pyhdfd78af_0'
-        : 'quay.io/biocontainers/savana:1.3.7--pyhdfd78af_0'}"
+        ? 'https://depot.galaxyproject.org/singularity/savana:1.3.8--pyhdfd78af_0'
+        : 'quay.io/biocontainers/savana:1.3.8--pyhdfd78af_0'}"

     input:
     tuple val(meta), path(tumour), path(tumour_index), path(normal), path(normal_index)
     tuple val(meta2), path(ref), path(ref_index)
+    tuple val(meta3), path(contigs)

     output:
     tuple val(meta), path("${prefix}.sv_breakpoints.vcf"), emit: sv_breakpoints_vcf
@@ -24,6 +25,7 @@
     script:
     def args = task.ext.args ?: ''
     prefix = task.ext.prefix ?: "${meta.id}"
+    def contigs_arg = contigs ? "--contigs ${contigs}" : ""
     """
     savana run \\
         --tumour ${tumour} \\
@@ -32,6 +34,7 @@
         --outdir "./outdir" \\
         --sample ${prefix} \\
         --threads ${task.cpus} \\
+        ${contigs_arg} \\
         ${args}

     mv ./outdir/* .

Changes in environment.yml file:
--- modules/nf-core/savana/run/environment.yml
+++ modules/nf-core/savana/run/environment.yml
@@ -4,4 +4,4 @@
 dependencies:
-  - "bioconda::savana=1.3.7"
+  - "bioconda::savana=1.3.8"

************************************************************
